### PR TITLE
Migrate parser to Chevrotain (feature-flagged)

### PR DESCRIPTION
**Pull request by kadraman**
#### Description:

- Adds a Chevrotain-based lexer/parser/transformer for the BeatBax language (feature-flag gated; legacy parser preserved).
- Includes token specs, lazy-loaded lexer, CST grammar, and CST→AST transformer that preserves existing AST shape.
- Adds comprehensive tests: lexer, parser, transformer, modifiers, edge cases, and parser parity tests comparing legacy vs Chevrotain parses.
- Adds an export-compare script to validate JSON/MIDI/UGE parity and a GitHub Actions workflow (.github/workflows/chevrotain-parity.yml) to run the matrix with/without the new parser.
- Updates docs/DEVNOTES with usage and testing instructions.

#### Included commits:

-- 1f26d41 (initial chevrotain implementation)
-- 8216bf1 (tests, parity, CI, docs)

### Checklist

- [ X] My changes add required tests and they pass.
- [ X] I ran `npm test` locally and all tests passed.
- [ X] I updated README or docs if required.
- [ X] This PR follows repository coding style and guidelines.

#### Files/areas to review:

- packages/engine/src/parser/chevrotain/*
- packages/engine/tests/chevrotain.ts and parser-parity.test.ts
- docs/features/chevrotain-migration.md
 
#### Testing notes:

- All engine tests pass locally.
- To run with Chevrotain enabled: PowerShell: `$env:BEATBAX_PARSER='chevrotain'; npm run engine:test`
- CI runs both default and `BEATBAX_PARSER=chevrotain` and executes artifact parity checks.

#### Backward compatibility & safety:

- Default parser unchanged; Chevrotain is enabled only via BEATBAX_PARSER.
- AST shape preserved to avoid breaking downstream consumers.

### CODE REVIEW

File: packages/engine/src/parser/chevrotain/parser.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        { ALT: () => this.SUBRULE(this.patStmt) },
+        { ALT: () => this.SUBRULE(this.instStmt) },
+        { ALT: () => this.SUBRULE(this.seqStmt) },
+        { ALT: () => this.SUBRULE(this.channelStmt) },
+      { ALT: () => this.SUBRULE(this.songStmt) },
+      { ALT: () => this.SUBRULE(this.exportStmt) },
+        { ALT: () => this.CONSUME(Chip) },
+        { ALT: () => this.CONSUME(Bpm) },
+        { ALT: () => this.CONSUME(Play) },
+        { ALT: () => this.CONSUME(Export) },
```
[the remark this answers is not in the document]
Comment (kadraman) [reply packages/engine/src/parser/chevrotain/parser.ts]: Implemented suggestion

File: packages/engine/src/parser/chevrotain/transformer.ts [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        const fmt = fmtTok.image;
+        let dest: string | undefined = undefined;
+        if (destTok) {
+          dest = destTok.image.replace(/^['"]|['"]$/g, '');
+        }
+        (ast.metadata.exports as any) = (ast.metadata.exports || []);
+        (ast.metadata.exports as any).push({ format: fmt, dest });
+      }
+
+      else if (key === 'channelStmt') {        const eqIdx = tokens.findIndex(t => t.tokenType && t.tokenType.name === 'Equals');
```
[the remark this answers is not in the document]
Comment (kadraman) [reply packages/engine/src/parser/chevrotain/transformer.ts]: Implemented suggestion

File: packages/engine/src/parser/chevrotain/parser.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+          ALT: () => {
+            this.CONSUME(Id);
+            this.OPTION(() => {
+              this.CONSUME(LParen);
+              this.OPTION(() => this.OR([
+                { ALT: () => this.CONSUME(NumberLiteral) },
+                { ALT: () => this.CONSUME(Id) },
+                { ALT: () => this.CONSUME(StringLiteral) },
+              ]));
+              this.OPTION(() => { this.CONSUME(Comma); this.CONSUME(NumberLiteral); });
```
[the remark this answers is not in the document]
Comment (kadraman) [reply packages/engine/src/parser/chevrotain/parser.ts]: Implemented suggestion

File: packages/engine/tests/parser-parity.test.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+      const a1 = legacyParse(src);
+      const res = await parseWithChevrotain(src);
+      if (res.errors && res.errors.length) throw new Error('Chevrotain parser errors: ' + JSON.stringify(res.errors));
+      const a2 = res.ast;
+      expect(a2).toBeTruthy();
+      const n1 = normalize(a1);
+      const n2 = normalize(a2);
+      expect(n2).toEqual(n1);
+    });
+  }
```
[the remark this answers is not in the document]
Comment (kadraman) [reply packages/engine/tests/parser-parity.test.ts]: Implemented suggestion

File: packages/engine/src/parser/chevrotain/tokens.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  { name: 'Seq', pattern: /seq/ },
+  { name: 'Channel', pattern: /channel/ },
+  { name: 'Chip', pattern: /chip/ },
+  { name: 'Song', pattern: /song/ },
+  { name: 'Bpm', pattern: /bpm/ },
+  { name: 'Play', pattern: /play/ },
+  { name: 'Export', pattern: /export/ },
+
+  // literals
+  { name: 'StringLiteral', pattern: /"(?:[^"\\]|\\.)*"|'(?:[^'\\]|\\.)*'/ },
```
[the remark this answers is not in the document]
Comment (kadraman) [reply packages/engine/src/parser/chevrotain/tokens.ts]: Implemented suggestion

File: packages/engine/scripts/compare-exports.mjs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    // JSON export compare
+    const out1 = path.join(outDir, `${base}.legacy.json`);
+    const out2 = path.join(outDir, `${base}.chev.json`);
+    await exportJSON(a1, out1, { debug: false });
+    await exportJSON(a2, out2, { debug: false });
+    const j1 = JSON.parse(fs.readFileSync(out1, 'utf8'));
+    const j2 = JSON.parse(fs.readFileSync(out2, 'utf8'));
+    // Normalize non-deterministic fields (exportedAt)
+    if (j1 && j1.song) delete j1.exportedAt;
+    if (j2 && j2.song) delete j2.exportedAt;
```
[the remark this answers is not in the document]
Comment (kadraman) [reply packages/engine/scripts/compare-exports.mjs]: Implemented suggestion
